### PR TITLE
8344943: Mark not subclassable classes final in java.base exported classes

### PR DESCRIPTION
**Pull request by eirbjo**
Please review this PR which adds the `final` modifier to non-subclassable classes in `java.base`.

The classes were identified using an automated analysis. See CSR for details.

Besides simply adding the `final` access modifier, the PR:

* Updates a note in `java.lang.constant.DynamicCallSiteDesc` to not reference subtypes. See CSR for discussion.
* Removes the class `java.lang.Runtime` from the test `test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSMs.java`
* Updates the copyright year of affected source files

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8345054](https://bugs.openjdk.org/browse/JDK-8345054) to be approved

### Issues
 * [JDK-8344943](https://bugs.openjdk.org/browse/JDK-8344943): Mark not subclassable classes final in java.base exported classes (**Enhancement** - P4)
 * [JDK-8345054](https://bugs.openjdk.org/browse/JDK-8345054): Mark not subclassable classes final in java.base exported classes (**CSR**)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**) Review applies to [d93dcf8b](https://git.openjdk.org/jdk/pull/22389/files/d93dcf8b45055f19f4235cb6b5e92273390392a1)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**) Review applies to [d93dcf8b](https://git.openjdk.org/jdk/pull/22389/files/d93dcf8b45055f19f4235cb6b5e92273390392a1)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22389/head:pull/22389` \
`$ git checkout pull/22389`

Update a local copy of the PR: \
`$ git checkout pull/22389` \
`$ git pull https://git.openjdk.org/jdk.git pull/22389/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22389`

View PR using the GUI difftool: \
`$ git pr show -t 22389`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22389.diff">https://git.openjdk.org/jdk/pull/22389.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22389#issuecomment-2596851719)
</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java [hunk 1]
```diff
@@ -41,12 +41,9 @@
  * A <a href="package-summary.html#nominal">nominal descriptor</a> for an
  * {@code invokedynamic} call site.
  *
- * <p>Concrete subtypes of {@linkplain DynamicCallSiteDesc} should be immutable
- * and their behavior should not rely on object identity.
```
Comment (liach) [inline src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java:45]: Please reword this to something like:
```
{@code DynamicCallSiteDesc} is immutable and its behavior does not rely on object identity.
```
This is given in `ConstantDesc`, but `DynamicCallSiteDesc` does not extend `ConstantDesc` so the removal is dubious.

**Review by liach — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java (hunk 1 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java:45]: Thanks, I've updated the PR and CSR to retain the note about immutability and object identity, removing the reference to subtypes:

```
This class is immutable and its behavior does not rely on object identity
```

I opted to replace the self-reference `{@code DynamicCallSiteDesc}` with just "This class" here. Let me know if you prefer spelling out the class name.

File: src/java.base/share/classes/java/lang/module/ModuleDescriptor.java [hunk 2]
```diff
@@ -1312,7 +1312,7 @@ private ModuleDescriptor(String name,
      * Creates a module descriptor from its components.
      * The arguments are pre-validated and sets are unmodifiable sets.
      */
-    ModuleDescriptor(String name,
+    private ModuleDescriptor(String name,
                      Version version,
                      Set<Modifier> modifiers,
                      Set<Requires> requires,
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:1318]: I assume you'll fix the alignment if this goes ahead.

Comment (eirbjo) [reply src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:1318]: Thanks! Yes, that's good feedback. There are some non-API changes that I'd like to make, such as this and updating the comment for the `InterfaceAddress` constructor.

I prefered to make the initial PR draft contain just the CSR changes, such that PR would be useful and clean as a Specification change, since the diff is a bit noisy.

I'll keep your review comment here unresolved for now and get back to it pending CSR approval.

File: src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java (hunk 1 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java:45]: For the record:

It was not completely clear to me whether restricting these two package-private constructors to private is considered a specification change in the CSR sense. I opted to point them out in the CSR, just to make the change abundantly clear.

File: src/java.base/share/classes/java/lang/module/ModuleDescriptor.java (hunk 2 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:1318]: For the record:

It was not completely clear to me whether restricting these two package-private constructors to private is considered a specification change in the CSR sense. I opted to point them out in the CSR, just to make the change abundantly clear.

File: src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java [hunk 3]
```diff
@@ -41,12 +41,11 @@
...
+ * <p>This class is immutable and its behavior does not rely on object identity
```
Comment (liach) [inline src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java:44]:
```suggestion
 * <p>A {@code DynamicCallSiteDesc} is immutable and its behavior does not
 * rely on object identity.
```
This describes an object, not a class, and please close sentences with a period.
(We use fragments for param and return block tags instead)

File: src/java.base/share/classes/java/lang/module/ModuleDescriptor.java (hunk 2 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:1318]: This constructor is part of the public API: package-private is not accessible, and it does not appear in the API docs: https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/module/ModuleDescriptor.html

File: src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java (hunk 3 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/lang/constant/DynamicCallSiteDesc.java]: Thanks, fixed in 6156b46 :)

**Comment by eirbjo** [comment]
Following some offline discussion with @liach, we decided to leave out the constructor access updates in `ModuleDescriptor` and `InterfaceAddress` for now. They are not considered a specification change and adds noise for the CSR review.

These changes may be revisited in this PR pending CSR approval, or can be addressed in follow-up PRs.

File: src/java.base/share/classes/java/net/InterfaceAddress.java [hunk 4]
```diff
@@ -45,7 +45,7 @@ public class InterfaceAddress {
      * Package private constructor. Can't be built directly, instances are
```
Comment (liach) [inline src/java.base/share/classes/java/net/InterfaceAddress.java]: The comment is outdated.  This constructor is actually called via JNI like `(*env)->NewObject(env, ni_ibcls, ni_ibctrID)` in `NetworkInterface.c`.  I think we need to consult a network engineer to decide what is the best way to comment on this constructor.

**Review by liach — APPROVED** [review]
Looks good for 25.

**Comment by eirbjo** [comment]
Now that the CSR for this change is "final-ly" approved, I'm marking this PR ready for review.

Since the CSR approval, the PR has merged with a recent master and the copyright headers are updated for 2025.

**Comment by seanjmullan** [comment]
@bradfordwetmore @wangweij please review the change to `DRBGParameters` and make sure we are ok with it.

**Comment by wangweij** [comment]
The `DRBGParameters` change looks good. It’s a utility class with static methods and a private constructor, so any attempt to extend it should result in a compile-time error.

**Review by wangweij — APPROVED** [review]

**Comment by eirbjo** [comment]
Thanks everyone for your patient reviews in this PR. I'll integrate this now.

/integrate
